src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3310,70 +3335,282 @@ new_readdir(DIR *dirp)
 	return &mydir->ents[mydir->num_ents];
 }
 
-/**
- *static char** pre_envp(char *const envp[])
- *{
- *	int	i, num_entry = 0;
- *	char	**new_envp;
- *
- *	if (envp == NULL) {
- *		num_entry = 0;
- *	} else if (envp[0] == NULL) {
- *		num_entry = 0;
- *	} else {
- *		while (envp[num_entry])	{
- *			num_entry++;
- *		}
- *	}
- *
- *	new_envp = malloc(sizeof(char *) * (num_entry + 5));
- *	assert(new_envp != NULL);
- *	for (i = 0; i < num_entry; i++)	{
- *		new_envp[i] = envp[i];
- *	}
- *	for (i = 0; i < 5; i++)	{
- *		new_envp[num_entry + i] = envp_lib[i];
- *	}
- *
- *	return new_envp;
- *}
- *
- *static int
- *new_execve(const char *filename, char *const argv[], char *const envp[])
- *{
- *	printf("DBG> execve(%s)\n", filename);
- *	return real_execve(filename, argv, pre_envp(envp));
- *}
- *
- *static int
- *new_execvp(const char *filename, char *const argv[])
- *{
- *	printf("DBG> execvp(%s)\n", filename);
- *	return real_execve(filename, argv, pre_envp(NULL));
- *}
- *
- *static int
- *new_execv(const char *filename, char *const argv[])
- *{
- *	printf("DBG> execv(%s)\n", filename);
- *	pre_envp(NULL);
- *	return real_execve(filename, argv, pre_envp(NULL));
- *}
- *
- *static pid_t
- *new_fork(void)
- *{
- *	pid_t pid;
- *	pid = real_fork();
- *	if (pid) {
- *		// parent process: do nothing
- *		return pid;
- *	} else {
- *		init_dfs();
- *		return pid;
- *	}
- *}
+extern char **__environ;
+
+/* This is the number of environmental variables that would be forced to set in child process.
+ * "LD_PRELOAD" is a special case and it is not included in the list.
+ */
+#define LEN_ENV_LIST (10)
+/* Including LEN_ENV_LIST and "LD_PRELOAD" */
+#define N_ENV_APPEND (LEN_ENV_LIST + 1)
+
+/* Environmental variables could be cleared in some applications. To make sure all libpil4dfs
+ * related env properly set, we intercept execve and its variants to check envp[] and append our
+ * env when needed.
+ * pre_envp() scans provided envp to look for libpil4dfs related envs (env_list[]) and returns a
+ * new envp with missing libpil4dfs related envs appended.
+ * The fd for daos logging may be not valid here, so printf() is used in this function.
  */
+static char **
+pre_envp(char *const envp[])
+{
+	int    i, j, rc;
+	int    len, len2, len_total;
+	char **new_envp;
+	char  *pil4df_path;
+	char  *new_preload_str;
+	int    num_entry                  = 0;
+	int    num_entry_found            = 0;
+	int    idx_preload                = -1;
+	bool   preload_included           = false;
+	bool   pil4dfs_in_preload         = false;
+
+	/* "LD_PRELOAD" is a special case which needs .*/
+	char   env_list[LEN_ENV_LIST][32] = {"D_IL_REPORT", "DAOS_MOUNT_POINT", "DAOS_POOL",
```

Comment:
Is the 32 or LEN_ENV_LIST required here?  As you're populating this list you can probably let the compiler do the work here.  See https://github.com/daos-stack/daos/blob/master/src/cart/crt_init.c#L21 for now to do this dynamically without using constants.
```suggestion
	char   *env_list[] = {"D_IL_REPORT", "DAOS_MOUNT_POINT", "DAOS_POOL",
```

Reply:
Fixed as you suggested. Thank you very much!